New post gui challenge split text (#5125)
GoogleChrome/web.dev: src/site/content/en/blog/building-split-text-animations/index.md

---
layout: post
title: Building split text animations
subhead: A foundational overview of how to build split letter and word animations.
authors:
  - adamargyle
description: A foundational overview of how to build split letter and word animations.
date: 2021-04-20
hero: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/XHfyVzz2S49XK7jRl9Xx.png
thumbnail: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/VgfJpMKN2a5p0dKRFxse.png
tags:
  - blog
  - css
  - dom
  - javascript
  - animation
---

> In this post I want to share thinking on ways to solve split text animations and
> interactions for the web that is minimal, accessible and works across browsers.

Reviewer: ```suggestion
In this post I want to share thinking on ways to solve split text animations and
interactions for the web that are minimal, accessible, and work across browsers.
```

Try the [demo](https://gui-challenges.web.app/split-text/dist/).

<figure class="w-figure w-figure--fullbleed">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/H9NpmghPMDoJzJLNjk2V.mp4",
    class="w-screenshot",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">
    <a href="https://gui-challenges.web.app/split-text/dist/">Demo</a>
  </figcaption>
</figure>

If you prefer video, here's a YouTube version of this post:

{% YouTube '3hvN7bkjZBk' %}

## Overview

Split text animations can be amazing. We'll be barely scratching the surface of
animation potential in this post, but it does provide a foundation to build
upon. The goal is to animate progressively. The text should be readable by

> default, with the animation built on top of it. Split text motion effects can

Reviewer: ```suggestion
default, with the animation built on top. Split text motion effects can
```

animation. If motion is OK, we go ahead and chop it up into pieces. Here's a
preview of the HTML after Javascript has split the text by letter. 

<figure class="w-figure">
  {% Img
    src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/prwDis2IMcZxFD39mnWF.png",
    alt="screenshot of the Chrome devtools with the Elements panel open and reduced motion set to 'reduce' and the h1 is shown unsplit",
    class="w-screenshot",
    width="800", height="517"
  %}
  <figcaption class="w-figcaption">
    User is OK with motion; text split into multiple <span> elements
  </figcaption>
</figure>

### Preparing motion conditionals

The conveniently [available](https://caniuse.com/prefers-reduced-motion) `@media
(prefers-reduced-motion: reduce)` media query will be used from CSS and
Javascript in this project. This media query is our primary conditional for
deciding to split text or not. The CSS media query will be used to withhold
transitions and animations, while the Javascript media query will be used to
withhold the HTML manipulation. 

{% Banner 'neutral' %} **Question:** What else should be used to withhold split
text animations?{% endBanner %}

#### Preparing the CSS conditional

> I used PostCSS to enable the syntax of Media Queries Level 5, where I can store

Reviewer: ```suggestion
I used PostCSS to enable the syntax of [Media Queries Level 5](https://drafts.csswg.org/mediaqueries-5/), where I can store
```

```

#### Preparing the JS conditional

In Javascript, the browser provides a way to check media queries, I used
[destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)
to extract and rename the boolean result from the media query check:

```js
const {matches:motionOK} = window.matchMedia(
  '(prefers-reduced-motion: no-preference)'
)
```

I can then test for `motionOK`, and only change the document if the user has not
requested to reduce motion.

```js
if (motionOK) {
  // document split manipulations
}
```

I can check the same value by using PostCSS to enable the `@nest` syntax from
[Nesting Draft 1](https://drafts.csswg.org/css-nesting-1/). This allows me to
store all the logic about the animation and it's style requirements for the
parent and children, in one place:

```css
letter-animation {
  @media (--motionOK) {
    /* animation styles */
  }
}
```

With the PostCSS custom property and a Javascript boolean, we're ready to
conditionally upgrade the effect. That rolls us into the next section where I
break down the Javascript for transforming strings into elements.

## Splitting Text

> Text letters, words, lines, etc cannot be individually animated with CSS or JS.

Reviewer: ```suggestion
Text letters, words, lines, etc., cannot be individually animated with CSS or JS.
```
